### PR TITLE
[ShogunBoard] Add parameter observable and observe_param() method.

### PR DESCRIPTION
**Pull request by geektoni**
The subscribe_to_parameters() method is still empty because I need to know
how `ParameterObserver` will be implemented @vigsterkr

### CODE REVIEW

**Comment by vigsterkr** [comment]
one way to define the interface is:
```
class ParameterObserver: public SGObject
{
public:
     ParameterObserver();
     virtual ~ParameterObserver();
     
     virtual void on_next(std::pair<std::string, Any>& ) = 0;
     virtual void on_error(std::exception_ptr) = 0;
     virtual void on_complete() = 0;
};
```
this way we avoid the template virtual hell :)

File: src/shogun/base/SGObject.cpp [hunk 1]
```diff
@@ -800,3 +806,17 @@ bool CSGObject::type_erased_has(const BaseTag& _tag) const
 {
 	return self->has(_tag);
 }
+
+void CSGObject::subscribe_to_parameters(ParameterObserverInterface* obs)
```
Comment (geektoni) [inline src/shogun/base/SGObject.cpp]: @vigsterkr I've added a simple `ParameterObserverInterface` class, as you suggested yesterday, to implement the behavior of `subscribe_to_parameters()`.

**Comment by geektoni** [comment]
@vigsterkr mmh it seems that some commits made with previous PRs disappeared from the shogun's branch feature/premature-stopping (that's why this PR is so crowded). Why did it happen? :/

**Comment by vigsterkr** [comment]
oooooooooooooooooooh fuck.... i think it might be my stupid mistake as i think i did the rebase on my local feature branch that was not synced with the latest upstream feature branch :((((

File: src/shogun/CMakeLists.txt [hunk 2]
```diff
@@ -286,6 +286,20 @@ ELSE()
     SHOGUN_INCLUDE_DIRS(SCOPE PUBLIC SYSTEM ${rxcpp_INCLUDE_DIR})
 ENDIF()
 
+# TFLogger package
+FIND_PACKAGE(TFLogger 0.1.0 CONFIG)
+IF (TFLogger_FOUND)
+    SHOGUN_INCLUDE_DIRS(SCOPE PUBLIC SYSTEM ${TFLogger_INCLUDE_DIR})
```
Comment (vigsterkr) [inline src/shogun/CMakeLists.txt]: do we need really public system? as this is really gonna be just statically linked into .so... so afaik PRIVATE should be enough, no?

File: src/shogun/CMakeLists.txt [hunk 3]
```diff
@@ -286,6 +286,20 @@ ELSE()
...
+    target_link_libraries(shogun PUBLIC tflogger::tflogger)
```
Comment (vigsterkr) [inline src/shogun/CMakeLists.txt]: same here i reckon a private would be enough...

**Comment by geektoni** [comment]
https://gist.github.com/geektoni/a32be8e4460215ea8bcd1ccf54ac4cb7
screenshot from 2017-06-30 17-23-12

File: src/shogun/lib/ParameterObserverScalar.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+void ParameterObserverScalar::on_complete()
+{
+}
+
+bool ParameterObserverScalar::filter(const std::string& param)
+{
+    if (m_parameters.size() == 0)
+        return true;
+
+	for (auto v : m_parameters)
```
Comment (vigsterkr) [inline src/shogun/lib/ParameterObserverScalar.cpp]: indentation ?

File: src/shogun/lib/ParameterObserverScalar.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    const std::string& filename, std::vector<std::string>& parameters)
+    : ParameterObserverInterface(filename, parameters)
+{
+    m_writer.init();
+}
+
+ParameterObserverScalar::~ParameterObserverScalar()
+{
+    m_writer.flush();
+	m_writer.close();
```
Comment (vigsterkr) [inline src/shogun/lib/ParameterObserverScalar.cpp]: indentation

File: src/shogun/lib/ParameterObserverInterface.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+	protected:
+		/**
+		 * List of parameter's names we want to monitor
+		 */
+		std::vector<std::string> m_parameters;
+		/**
+		 * Writer object which will be used to write tensorflow::Event files
+		 */
+		tflogger::TensorFlowLogger m_writer;
```
Comment (vigsterkr) [inline src/shogun/lib/ParameterObserverInterface.h]: mmm i'm not so sure if an interface (abstract class) should actually contain this

File: src/shogun/lib/tfhistogram/histogram.cpp [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "histogram.h"
+#include <float.h>
+#include <math.h>
+#include <vector>
+#include <tflogger/summary.pb.h>
```
Comment (vigsterkr) [inline src/shogun/lib/tfhistogram/histogram.cpp]: this implementation and the whole tfhistogram needs guard of HAVE_TFLOGGER

**Comment by vigsterkr** [comment]
mmm seems the changes really affect the compile time, i'm merging and we'll try to figure out in the feature branch how to solve this...

**Pull request merged**